### PR TITLE
Heatmap clustering

### PR DESCRIPTION
**Pull request by mhuyck**
This update implements heatmap clustering and, since node clustering takes a while, restores status updates for the user (temporarily removed when code was reorganized to cache activity data) so they understand that the browser is not freezing up during lengthy operations.

### CODE REVIEW

**Comment by mhuyck** [comment]
@dongbohu, @rzelayafavila: Just noticed the new GitHub "reviewers" feature, so moved you from "Assignees" to the Reviewers section.

File: interface/package.json [hunk 1]
```diff
@@ -14,7 +14,9 @@
   "engines": {
     "node": ">=4.0"
   },
-  "dependencies": {},
+  "dependencies": {
+    "hclusterjs": "^1.3.0"
```
Comment (dongbohu) [inline interface/package.json]: It is possible to specify a git repository in **bower.json** for installation. I am not sure whether we should do it to have hclusterjs package in `vendors/` directory instead of in `node_modules/`.

Comment (dongbohu) [inline interface/package.json]: https://github.com/bower/spec/blob/master/json.md#dependencies

File: interface/src/app/analyze/analysis/analysis.js [hunk 2]
```diff
@@ -187,21 +221,20 @@ AnnotationType, SampleBin) {
...
   // FIXME implement this loop as a method on SampleBin? (prob. with caching)
+  $scope.analysis.status = 'Retrieving sample details';
+  var pArrSamples = [];
   for (var i = 0; i < SampleBin.heatmapData.samples.length; i++) {
-    SampleBin.getSampleDetails(SampleBin.heatmapData.samples[i]);
+    pArrSamples.push(
+      SampleBin.getSampleDetails(SampleBin.heatmapData.samples[i])
+    );
   }
+  $q.all(pArrSamples).then(function() {
```
Comment (dongbohu) [inline interface/src/app/analyze/analysis/analysis.js]: I like your way of using `$q.all()` to handle the promises. On the other hand, do you think it will be worth the time to add a new API on the backend to retrieve all data of samples that are related to a certain node? Then the frontend will be simpler. (See my latest PR, which adds a new API to get the data of all samples that are related to a given experiment.)

File: interface/src/app/analyze/analysis/sampleBin.js [hunk 3]
```diff
@@ -75,48 +78,116 @@ function($log, $cacheFactory, $q, Sample, Activity) {
...
+      // reformat data from heatmapData.activity to a form that can be used
+      // by hcluster.js: need a separate array of objects for each node
+      var firstSampleNodes = this.activityCache.get(
+        this.heatmapData.samples[0]
+      );
+      var retval = firstSampleNodes.map(function(val, i, arr) {
+        var nodeObject = {
+          'id': val.node
+        };
+        nodeObject.activity = this.heatmapData.samples.map(
```
Comment (dongbohu) [inline interface/src/app/analyze/analysis/sampleBin.js]: Can you add more comments here? I found the map inside map hard to follow.

File: interface/src/common/statusBar/statusBar.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+angular.module('statusBar', [])
+
+.directive('statusBar', function() {
+  return {
+    restrict: 'E',
+    scope: {
+      msgText: '='
```
Comment (dongbohu) [inline interface/src/common/statusBar/statusBar.js]: I like this `<staus-bar>` gadget. Since **msgText** is read-only, we can use **@** instead of **=** here (then you have change the html above into:
```
<status-bar msgText="{{sample.status}}"></status-bar>
```

Another way is use the new **component()** function, which was introduced since Angular 1.5 for simple directives.
https://docs.angularjs.org/guide/component
(I am not saying that **component** is better than **directive** here, but I think it is good to know this option.)

File: interface/src/common/statusBar/statusBar.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+angular.module('statusBar', [])
```
Comment (mhuyck) [inline interface/src/common/statusBar/statusBar.js]: Right: I realized we had several places that could benefit from having a "status bar" done in a standardized way, but it's not adage-server specific, so I put it into `interface/src/common`.

The camelCase naming follows the convention throughout ngBoilerplate of using camelCase in JavaScript source and hyphenated-names in HTML. I have done nothing special to make the translation between the two work, so this appears to be a feature of Angular and, I assume, the preferred way to do it. Please correct me if I'm wrong!

File: interface/src/app/analyze/sample/sampleDetail.tpl.html [hunk 6]
```diff
@@ -10,10 +10,7 @@
     sample.related_experiments[0].accession, sample.results.ml_data_source)
   }}">{{sample.results.ml_data_source}}</a></small>
 </h3>
-<p ng-if="sample.status" class="alert alert-warning" role="alert">
-  Status: {{sample.status}}<br>
-  If this message persists, please try refreshing your browser.
-</p>
+<status-bar msg-text="sample.status"></status-bar>
```
Comment (mhuyck) [inline interface/src/app/analyze/sample/sampleDetail.tpl.html]: Thank you! If there are other places that can use this, we should probably standardize across our codebase.

File: interface/package.json (hunk 1 above)
Comment (mhuyck) [inline interface/package.json]: I believe this is the first NPM module we are deploying into our front-end code rather than using as a tool operating in our build process. I think this means it belongs in `dependencies` rather than `dev-dependencies`.

Would there be some benefit from changing things around to have it appear in `vendors/` instead?

File: interface/src/app/analyze/analysis/analysis.js (hunk 2 above)
Comment (mhuyck) [inline interface/src/app/analyze/analysis/analysis.js]: I thought about this for a while. Retrieving samples independently allows us to do caching at the level of samples. This means that if a user adds or removes a sample from the heatmap we can rebuild it much more quickly than we could otherwise.

In addition, although I haven't implemented it yet, we can ultimately begin fetching the analysis data into the cache as soon as an individual sample is added to the `sampleBin` (rather than waiting until the `analysis` view is opened, as we do now). This will make the apparent performance of the UI even snappier.

No need for a new API, by the way. The `Activity` API can already filter on `sample_id__in`. Before I implemented the cache this is the way it was done, in fact. Without caching, it *is* more efficient to get all of the activity data in a single call, but with the cache, I think this way is better.

Does that approach make sense?

File: interface/src/common/statusBar/statusBar.js (hunk 4 above)
Comment (mhuyck) [inline interface/src/common/statusBar/statusBar.js]: I didn't know about `component()` yet... thank you for pointing it out! I'll have a look and see if it makes sense. If not, I agree with your idea of using `@` rather than `=`.

- [ ] look into using `component()` here

File: interface/src/app/analyze/analysis/analysis.js (hunk 2 above)
Comment (mhuyck) [inline interface/src/app/analyze/analysis/analysis.js]: Whoops. I should have looked more closely at the context here... The relevant API is `Sample`, not `Activity`. But the logic I'm getting at is the same. Caching is not yet implemented for this `Sample` data, but once I have generalized that code, the UI will cache this sample data as well as the activity data so the same logic applies.

File: interface/src/common/statusBar/statusBar.js (hunk 5 above)
Comment (mhuyck) [inline interface/src/common/statusBar/statusBar.js]: As usual, StackOverflow provides a good explanation. Apparently the hyphenated-names are sometimes called kebab-case. LOL!

http://stackoverflow.com/questions/33460274/angularjs-directive-naming-conventions

Where I probably saw this was in the Angular tutorial here: https://docs.angularjs.org/tutorial/step_00#what-is-the-code-doing-

File: interface/package.json (hunk 1 above)
Comment (dongbohu) [inline interface/package.json]: There is probably no any other benefits except the consistency. I am okay with hosting this js library in `node_modules/`.

File: interface/src/app/analyze/analysis/analysis.js (hunk 2 above)
Comment (dongbohu) [inline interface/src/app/analyze/analysis/analysis.js]: Sure, thank you for your explanation.

File: interface/src/common/statusBar/statusBar.js (hunk 4 above)
Comment (dongbohu) [inline interface/src/common/statusBar/statusBar.js]: If you don't have time now, don't worry about the option of `component()` function. I am reading Angular's official doc recently and feel that it is an interesting feature to share with you guys.

File: interface/package.json (hunk 1 above)
Comment (mhuyck) [inline interface/package.json]: Okay, sounds good. Keep in mind that the `vendors/` and `node_modules/` directories exist only for the build and are not used by the deployed code. The minifying step used for the final "packaging" of the files will merge this file along with the others pulled from `vendors/` into our own customize `adage-ver#.js` file (currently `interface/bin/assets/adage-0.2.1a.js`) so, in the end, this file does actually get treated the same as the others in `vendors/`.

**Comment by dongbohu** [comment]
@mhuyck: Please note that the PR that I merged yesterday updated **build.config.js** too. You may want to update yours before merging. Otherwise you will have to resolve it later.

**Comment by mhuyck** [comment]
Thank you for the heads up @dongbohu! I have made it a practice to do a rebase just before merging, so I'll do that here too.

File: interface/src/common/statusBar/statusBar.js (hunk 4 above)
Comment (mhuyck) [inline interface/src/common/statusBar/statusBar.js]: The component approach is totally the way to go here. I don't want to drag out this PR right now, though, so I'm going to track this as an issue. There are a bunch of other places that could benefit from this simplification in our code, too, so I'm going to suggest we tackle those all at once.

Thank you so much for calling my attention to that feature of Angular, @dongbohu!

**Comment by dongbohu** [comment]
@mhuyck: I accidentally sent my last comment under the new issue that you created. In case you didn't see it, here it is:

Sounds good to me. I only have two minor concerns about this PR:
(1) It may be better to use @ in the scope (instead of =) when possible.
(2) Add some comments on the map inside map function calls.
After you address them, please feel free to merge it.

File: interface/src/common/statusBar/statusBar.js (hunk 4 above)
Comment (mhuyck) [inline interface/src/common/statusBar/statusBar.js]: I just tried converting from `msgText: '='` to use `'@'` instead. Whether or not I also make the corresponding change you mention to the HTML, the changes break the whole `statusBar`. I am not sure why yet, but the message text does not appear (the uninterpreted variable name appears instead) and the status bar does not disappear when it should either.

I don't have time to debug this now, so I'm reverting to using `'='` since that works and I'm making a note to follow up in the issue I just created.

**Comment by mhuyck** [comment]
@dongbohu: see the updates I've pushed for (2). I attempted the approach you suggested in (1) but it broke the `statusBar` so I reverted it. I've made a note to follow up on that in the issue about converting to Component.

**Comment by mhuyck** [comment]
@dongbohu Please let me know if my updated comments on the nested Array.prototype.map() code help clarify things.
